cloud-sql/mysql/servlet/src/main/webapp/index.jsp

```diff
@@ -0,0 +1,117 @@
+<%@ page import="java.util.List" %>
+<%@ page import="com.example.cloudsql.Vote" %>
+<!--
+Copyright 2018 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html lang="en">
+<head>
+    <title>Tabs VS Spaces</title>
+    <link rel="stylesheet"
+          href="https://cdnjs.cloudflare.com/ajax/libs/materialize/1.0.0/css/materialize.min.css">
+    <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/materialize/1.0.0/js/materialize.min.js"></script>
+</head>
+<body>
+<nav class="red lighten-1">
+    <div class="nav-wrapper">
+        <a href="#" class="brand-logo center">Tabs VS Spaces</a>
+    </div>
+</nav>
+
+<%
+    Integer tabVoteCt = (Integer) request.getAttribute("tabVoteCt");
+    Integer spaceVoteCt = (Integer) request.getAttribute("spaceVoteCt");
+    List<Vote> recentVotes = (List<Vote>) request.getAttribute("recentVotes");
+    int voteDiff = 0;
+    String leadTeam = "";
+    if (!tabVoteCt.equals(spaceVoteCt)) {
+      if (tabVoteCt > spaceVoteCt) {
+        leadTeam = "TABS";
+        voteDiff = tabVoteCt - spaceVoteCt;
+      } else {
+        leadTeam = "SPACES";
+        voteDiff = spaceVoteCt - tabVoteCt;
+      }
+    }
+%>
+<div class="section">
+    <div class="center">
+        <h4>
+            <% if(voteDiff != 0) { %>
```

Comment:
Could you provide more context on this? When should I use a taglib over the `<= ... %>`? Is `<%= ... &>` unsafe or inefficient? 

Is the advised taglib to use the JSTL core library?

Reply:
JSTL core library works. The taglibs are idiomatic in the JSP world, and you don't need to use `request.getAttribute()` to get the variables, since taglibs will be able to get the data directly from the attribute. It'll be cleaner, less code, and safer since it'll escape tags so you don't end up w/ script injections.

Just for example (and probably don't work as is since it's been a while since I had to write JSP :):

```
<c:choose>
<c:when test="${tabVoteCt == spaceVoteCt}">
  TABS and SPACES are tied
</c:when>
<c:when test="${tabVoteCt > spaceVoteCt}">
  TABS won
</c:when>
<c:when test="${tabVoteCt < spaceVoteCt}">
  SPACES won
</c:when>

</c:choose>
```

To iterate through the list:
```
<c:forEach items="${recentVotes}" var="vote">
  <li class="collection-item avatar">
    <c:choose>
      ...
    </c:choose>
  </li>
</c:forEach>
```

Reply:
Thanks for the background - I've made the adjustments.